lib/semantics/check-do-concurrent.cc

```diff
@@ -0,0 +1,254 @@
+        "RETURN not allowed in DO CONCURRENT"_err_en_US);
+  }
+  // C1137
+  void NoImageControl() {
+    messages_.Say(currentStatementSourcePosition_,
+        "image control statement not allowed in DO CONCURRENT"_err_en_US);
+  }
+  void Post(const parser::SyncAllStmt &) { NoImageControl(); }
+  void Post(const parser::SyncImagesStmt &) { NoImageControl(); }
+  void Post(const parser::SyncMemoryStmt &) { NoImageControl(); }
+  void Post(const parser::SyncTeamStmt &) { NoImageControl(); }
+  void Post(const parser::ChangeTeamConstruct &) { NoImageControl(); }
+  void Post(const parser::CriticalConstruct &) { NoImageControl(); }
+  void Post(const parser::EventPostStmt &) { NoImageControl(); }
+  void Post(const parser::EventWaitStmt &) { NoImageControl(); }
+  void Post(const parser::FormTeamStmt &) { NoImageControl(); }
+  void Post(const parser::LockStmt &) { NoImageControl(); }
+  void Post(const parser::UnlockStmt &) { NoImageControl(); }
+  void Post(const parser::StopStmt &) { NoImageControl(); }
+  void Post(const parser::EndProgramStmt &) { NoImageControl(); }
+
+  void Post(const parser::AllocateStmt &) {
+    if (anyObjectIsCoarray()) {
+      messages_.Say(currentStatementSourcePosition_,
+          "ALLOCATE coarray not allowed in DO CONCURRENT"_err_en_US);
+    }
+  }
+  void Post(const parser::DeallocateStmt &) {
+    if (anyObjectIsCoarray()) {
+      messages_.Say(currentStatementSourcePosition_,
+          "DEALLOCATE coarray not allowed in DO CONCURRENT"_err_en_US);
+    }
+    // C1140: deallocation of polymorphic objects
+    if (anyObjectIsPolymorphic()) {
+      messages_.Say(currentStatementSourcePosition_,
+          "DEALLOCATE polymorphic object(s) not allowed"
+          " in DO CONCURRENT"_err_en_US);
+    }
+  }
+  template<typename T> void Post(const parser::Statement<T> &) {
+    if (EndTDeallocatesCoarray()) {
+      messages_.Say(currentStatementSourcePosition_,
+          "implicit deallocation of coarray not allowed"
+          " in DO CONCURRENT"_err_en_US);
+    }
+  }
+  // C1141: cannot call ieee_get_flag, ieee_[gs]et_halting_mode
+  void Post(const parser::ProcedureDesignator &procedureDesignator) {
+    if (auto *name{std::get_if<parser::Name>(&procedureDesignator.u)}) {
+      // C1137: call move_alloc with coarray arguments
+      if (name->ToString() == "move_alloc"s) {
+        if (anyObjectIsCoarray()) {
+          messages_.Say(currentStatementSourcePosition_,
+              "call to MOVE_ALLOC intrinsic in DO CONCURRENT with coarray"
+              " argument(s) not allowed"_err_en_US);
+        }
+      }
+      // C1139: call to impure procedure
+      if (name->symbol && !isPure(name->symbol->attrs())) {
+        messages_.Say(currentStatementSourcePosition_,
+            "call to impure subroutine in DO CONCURRENT not allowed"_err_en_US);
+      }
+      if (name->symbol && fromScope(*name->symbol, "ieee_exceptions"s)) {
+        if (name->ToString() == "ieee_get_flag"s) {
```

Comment:
Is there a way we can not expose ToString() very often?  Perhaps name->Equals("") or something?  I don't have a good reason for hiding string other than just thinking my future self might not want to use String.

Reply:
All I see is ToString().

Reply:
You could cast the string to a CharBlock and then compare the CharBlocks, or extend CharBlock's comparison operators to accept string operands.

Reply:
`name->source == "ieee_get_flag"` should work now, right?

Reply:
Oh, sure, right.

Reply:
Thanks for the pointer. I'll change ToString(), etc. here.